device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,456 @@
+     * <p>
+     * If the multiplexing client is already open, then these device clients will automatically
+     * be opened, too. If the multiplexing client is not open yet, then these device clients will not be opened until
+     * {@link MultiplexingClient#open()} is called.
+     * <p>
+     * If the multiplexed connection is already open, then this call will asynchronously add each device client to the
+     * multiplexed connection, and then will block until all registrations have been completed.
+     * <p>
+     * Any proxy settings set to the provided device clients will be overwritten by the proxy settings of this multiplexing client.
+     * <p>
+     * The registered device clients must use the same transport protocol (AMQPS or AMQPS_WS) that this multiplexing client uses.
+     * <p>
+     * Each registered device client may have its own retry policy and its own SAS token expiry time, separate from every other registered device client.
+     * <p>
+     * The registered device clients must use symmetric key based authentication.
+     * <p>
+     * The registered device clients must belong to the same IoT Hub as all previously registered device clients.
+     * <p>
+     * If any of these device clients are already registered to this multiplexing client, then then this method will
+     * not do anything to that particular device client. All other provided device clients will still be registered though.
+     * <p>
+     * @throws InterruptedException If the thread gets interrupted while waiting for the registration to succeed. This
+     * will never be thrown if the multiplexing client is not open yet.
+     * @param deviceClients The device clients to associate with this multiplexing client.
+     */
+    public void registerDeviceClients(Iterable<DeviceClient> deviceClients) throws InterruptedException {
+        Objects.requireNonNull(deviceClients);
+
+        synchronized (this.operationLock)
+        {
+            List<DeviceClientConfig> deviceClientConfigsToRegister = new ArrayList<>();
+            for (DeviceClient deviceClientToRegister : deviceClients)
+            {
+                DeviceClientConfig configToAdd = deviceClientToRegister.getConfig();
+
+                // Overwrite the sslContext of the new client to match the multiplexing client
+                // settings if it was set. If it wasn't set, defer to the default client to set this ssl context
+                if (this.sslContext != null)
+                {
+                    configToAdd.getAuthenticationProvider().setSSLContext(this.sslContext);
+                }
+
+                // Overwrite the proxy settings of the new client to match the multiplexing client settings
+                configToAdd.setProxy(this.proxySettings);
+
+                boolean deviceAlreadyRegistered = false;
+                for (DeviceClient currentClient : deviceClientList)
```

Comment:
if we perform the registration operation again for a device that is already registered, will there be an issue? I am thinking abut how heavy this for loop will be, for ~1000 devices, and if making registration idempotent is an option.

Reply:
Good point. I can make this list a set instead to make this more idempotent

Reply:
HashSet's in Java are basically O(1) to add, so that will be a nice performance boost